File: scala-pure-io/src/main/scala/pureio/kata/guessnumber/GuessNumberApp.scala
package pureio.kata.guessnumber

import zio._
import zio.clock.Clock
import zio.console.Console
import zio.random.Random

import scala.util._

object GuessNumberApp extends App {
  override def run(args: List[String]): ZIO[Random with Console, Nothing, ExitCode] = game.as(ExitCode.success)

  def randomNumber: ZIO[Random, Nothing, Int] =
    random.nextIntBetween(1, 20)

  def getGuess: ZIO[Console, Nothing, Int] =
    getIntBetween(1, 20).retry(Schedule.forever).provideSomeLayer[Console](Clock.live).orDie

  def game: ZIO[Console with Random, Nothing, Unit] = for {
    _ <- console.putStrLn("Guess a number between 1 and 20.")
    number <- randomNumber
    _ <- guessLoop(number, 1)
  } yield ()

  def guessLoop(number: Int, attempt: Int): ZIO[Console, Nothing, Unit] = for {
    _ <- console.putStr(s"Attempt $attempt > ")
    guess <- getGuess

    won <-
      if (guess > number)
        console.putStrLn("It's too large.").as(false)
      else if (guess < number)
        console.putStrLn("It's too small.").as(false)
      else
        console.putStrLn(s"You won after $attempt attempt(s).").as(true)

    _ <- if (won) ZIO.unit else guessLoop(number, attempt + 1)
  } yield ()

  def parseInt(s: String): Either[NumberFormatException, Int] =
    Try(s.toInt).toEither
      .left.map {
        case ex: NumberFormatException => ex
        case ex => throw ex
      }

  val getInt: ZIO[Console, NumberFormatException, Int] =
    console.getStrLn.orDie
      .flatMap(s => ZIO.fromEither(parseInt(s)))

  def getIntBetween(min: Int, max: Int): ZIO[Console, NumberFormatException, Int] =
    getInt.filterOrElse(n => min <= n && n <= max)(n => ZIO.fail(new NumberFormatException(s"Number not between $min and $max ($n)")))
}

File: java-pure-io/src/main/java/pureio/kata/guessnumber/GuessNumberApp.java
package pureio.kata.guessnumber;

import io.vavr.control.Option;
import io.vavr.control.Try;
import pureio.presentation.console.purefree.Program;
import pureio.presentation.console.purefree.Unit;

import static pureio.kata.guessnumber.Random.randomBetween;
import static pureio.presentation.console.purefree.Console.getStrLn;
import static pureio.presentation.console.purefree.Console.putStrLn;

@SuppressWarnings({"CodeBlock2Expr", "Convert2MethodRef", "WeakerAccess"})
public class GuessNumberApp {
    public static void main(final String[] args) {
        Program.unsafeRun(mainApp);
    }

    public static final Program<Unit> mainApp =
            putStrLn("Guess a number between 1 and 20.").thenChain(__ -> {
                return randomBetween(1, 20).thenChain(number -> {
                    return guessLoop(number, 1);
                });
            });

    public static Program<Unit> guessLoop(final int number, final int attempt) {
        return putStrLn(String.format("Attempt %d>", attempt)).thenChain(__ -> {
            return getIntBetween(1, 20).thenChain(guess -> {
                return checkAttempt(number, attempt, guess).thenChain(won -> {
                    return won ? Program.yield(Unit.of()) : guessLoop(number, attempt + 1);
                });
            });
        });
    }

    public static Program<Boolean> checkAttempt(final int number, final int attempt, final int guess) {
        if (guess < number) {
            return putStrLn("It's too small.").thenTransform(__ -> false);
        } else if (guess > number) {
            return putStrLn("It's too large.").thenTransform(__ -> false);
        } else {
            return putStrLn(String.format("You won after %d attempt(s).", attempt)).thenTransform(__ -> true);
        }
    }

    public static Option<Integer> parseInt(final String s) {
        return Try.of(() -> Integer.valueOf(s)).toOption();
    }

    public static Program<Integer> getInt() {
        return getStrLn()
                .thenTransform(s -> parseInt(s))
                .thenChain(maybeInt -> {
                    return maybeInt.isDefined() ? Program.yield(maybeInt.get()) : /* RECURSE */ getInt();
                });
    }

    public static Program<Integer> getIntBetween(final int min, final int max) {
        return getInt().thenChain(i -> {
            return min <= i && i <= max ? Program.yield(i) : /* RECURSE */ getIntBetween(min, max);
        });
    }
}
